app/src/main/java/android/example/wordlehelper/Game.kt

```diff
@@ -1,162 +1,358 @@
+                val comparedLetter = (currentFocus?.parent as LinearLayout).getChildAt(index) as TextView
+                comparedLetter.setBackgroundColor(colorStates.black)
+            }
+        }
+    }
 
-                    MyMethods().enterPress(
-                        currentFocus as TextView,
-                        keyboardButton,
-                        wordList,
-                        goalWord,
-                        resources,
-                        keyboardLayout,
-                        this,
-                        this
-                    )
-                }
+    private fun applyColorToLetter(comparedLetter: TextView, color: Int) {
+        comparedLetter.setBackgroundColor(color)
+        val keyboardLayout = findViewById<LinearLayout>(R.id.keyboardContainerLayout)
+        keyboardLayout.forEach { keyboardRow ->
+            (keyboardRow as LinearLayout).forEach { keyboardButton ->
+                keyboardButton as Button
+                if (keyboardButton.text.toString() == comparedLetter.text.toString())
+                    keyboardButton.setBackgroundColor(color)
             }
         }
     }
+
+    private fun showEndGameDialog(context: Context, goalWord: String, game: Activity, messageText: String) {
+        MaterialAlertDialogBuilder(context)
+            .setMessage(messageText)
+            .setPositiveButton("Atrás") { _, _ ->
+            }
+            .setNegativeButton("Definición RAE de ${goalWord.uppercase()}") { _, _ ->
+                // Respond to positive button press
+                val website = Intent(
+                    Intent.ACTION_VIEW,
+                    Uri.parse("https://dle.rae.es/${goalWord}")
+                )
+                context.startActivity(website)
+            }
+            .setNeutralButton("Nueva Partida") { _, _ ->
+                // Respond to positive button press
+                game.finish()
+                val intent = Intent(game, game::class.java)
+                game.startActivity(intent)
+            }
+            .show()
+    }
+
+    private fun giveUpConfirmDialog(goalWord: String, context: Context, game: Game, statsManager: StatsManager) {
+        MaterialAlertDialogBuilder(context)
+            .setMessage("Seguro que quieres abandonar?")
+            .setNegativeButton("Cancelar") { _, _ -> }
+            .setPositiveButton("Abandonar") { _, _ ->
+                statsManager.onGiveUpUpdateStats()
+                showGiveUpDialog(goalWord, context, game)
+            }
+            .show()
+    }
+
+    private fun showGiveUpDialog(goalWord: String, context: Context, game: Game) {
+        MaterialAlertDialogBuilder(context)
+            .setMessage("La palabra era ${goalWord.uppercase()}")
+            .setNegativeButton("Salir a Menu Principal") { _, _ ->
+                val intent = Intent(context, MainActivity::class.java)
+                game.startActivity(intent)
```

Comment:
Should I just finish the game and not start MainActivity? I've noticed that when I do this, if I press the back button while on the Main Menu it will close that activity but reload the one created originally.